.emacs.d/configuration.org
#+TITLE: Emacs Configuration
#+AUTHOR: Josh Bairstow
#+EMAIL: joshbairstow@gmail.com

* Use sensible-defaults.el

Use [[https://github.com/hrs/sensible-defaults.el][sensible-defaults.el]] for some basic settings.

#+BEGIN_SRC emacs-lisp
  (load-file "~/git/sensible-defaults.el/sensible-defaults.el")
  (sensible-defaults/use-all-settings)
  (sensible-defaults/use-all-keybindings)
  (sensible-defaults/backup-to-temp-directory)
#+END_SRC

* =evil-mode=

Use =evil=.

#+BEGIN_SRC emacs-lisp
  (add-to-list 'load-path "~/git/evil")
  (require 'evil)
  (evil-mode 1)
#+END_SRC

Make adjustment to expression evaluation when the point is on a lisp closing paren in evil mode

#+BEGIN_SRC emacs-lisp
;;  (add-to-list 'load-path "~/git/evil-adjust")
;;  (require 'evil-adjust)
;;  (evil-adjust)
;; (evil-move-beyond-eol t)
#+END_SRC

Enable =surround= everywhere.

#+BEGIN_SRC emacs-lisp
;;  (global-evil-surround-mode 1)
#+END_SRC

Bind =C-p= to fuzzy-finding files in the current project.

#+BEGIN_SRC emacs-lisp
  (define-key evil-normal-state-map (kbd "C-p") 'projectile-find-file)
#+END_SRC

* UI preferences
** Tweak window chrome

I don't usually use the menu or scroll bar, and they take up useful space.

#+BEGIN_SRC emacs-lisp
  (tool-bar-mode 0)
  (menu-bar-mode 0)
  (when window-system
    (scroll-bar-mode -1))
#+END_SRC

** Update colour theme

#+BEGIN_SRC emacs-lisp
  (add-to-list 'custom-theme-load-path "~/.emacs.d/themes/")
#+END_SRC

The default frame title isn't useful. This binds it to the name of the current
project:

#+BEGIN_SRC emacs-lisp
  (setq frame-title-format '((:eval (projectile-project-name))))
#+END_SRC

** Use fancy lambdas

Why not?

#+BEGIN_SRC emacs-lisp
  (global-prettify-symbols-mode t)
#+END_SRC

** Highlight the current line

=global-hl-line-mode= softly highlights the background color of the line
containing point. It makes it a bit easier to find point, and it's useful when
pairing or presenting code.

#+BEGIN_SRC emacs-lisp
  (when window-system
    (global-hl-line-mode))
#+END_SRC

* Programming customizations

I like shallow indentation, but tabs are displayed as 8 characters by default.
This reduces that.

#+BEGIN_SRC emacs-lisp
  (setq-default tab-width 2)
#+END_SRC

#+BEGIN_SRC emacs-lisp
  (custom-set-variables
  ;; custom-set-variables was added by Custom.
  ;; If you edit it by hand, you could mess it up, so be careful.
  ;; Your init file should contain only one such instance.
  ;; If there is more than one, they won't work right.
  '(ansi-color-faces-vector
  [default default default italic underline success warning error])
  '(ansi-color-names-vector
  ["#242424" "#e5786d" "#95e454" "#cae682" "#8ac6f2" "#333366" "#ccaa8f" "#f6f3e8"])
  '(custom-enabled-themes (quote (tsdh-dark))))
  (custom-set-faces
  ;; custom-set-faces was added by Custom.
  ;; If you edit it by hand, you could mess it up, so be careful.
  ;; Your init file should contain only one such instance.
  ;; If there is more than one, they won't work right.
  )
#+END_SRC

#+BEGIN_SRC emacs-lisp
(setq scheme-root "Applications/MIT-Scheme.app/Contents/Resources")

;; (setq scheme-program-name
;;       (concat
;;        scheme-root "/mit-scheme "
;;        "--library " scheme-root " "
;;        "--band " scheme-root "/all.com "
;;        "-heap 10000"))

(setq scheme-program-name "/Applications/MIT-Scheme.app/Contents/Resources/mit-scheme")

(add-to-list 'load-path "~/git/rainbow-delimiters")
(require 'rainbow-delimiters)
(add-hook 'prog-mode-hook 'rainbow-delimiters-mode)
#+END_SRC

#+BEGIN_SRC emacs-lisp
;;; I prefer cmd key for meta
(setq mac-option-key-is-meta nil
      mac-command-key-is-meta t
      mac-command-modifier 'meta
      mac-option-modifier 'none)
(set-keyboard-coding-system nil)
#+END_SRC

#+BEGIN_SRC emacs-lisp
;;; Enable Ido mode by default
(setq ido-enable-flex-matching t)
(setq ido-everywhere t)
(setq ido-create-new-buffer 'always)
(ido-mode 1)
#+END_SRC

#+BEGIN_SRC emacs-lisp
;;; Setup mit-scheme config
(setq scheme-program-name "/usr/local/bin/mit-scheme")
#+END_SRC


.emacs.d/configuration.el

(load-file "~/git/sensible-defaults.el/sensible-defaults.el")
(sensible-defaults/use-all-settings)
(sensible-defaults/use-all-keybindings)
(sensible-defaults/backup-to-temp-directory)

(add-to-list 'load-path "~/git/evil")
(require 'evil)
(evil-mode 1)

;;  (add-to-list 'load-path "~/git/evil-adjust")
;;  (require 'evil-adjust)
;;  (evil-adjust)
;; (evil-move-beyond-eol t)

;;  (global-evil-surround-mode 1)

(define-key evil-normal-state-map (kbd "C-p") 'projectile-find-file)

(tool-bar-mode 0)
(menu-bar-mode 0)
(when window-system
  (scroll-bar-mode -1))

(add-to-list 'custom-theme-load-path "~/.emacs.d/themes/")

(setq frame-title-format '((:eval (projectile-project-name))))

(global-prettify-symbols-mode t)

(when window-system
  (global-hl-line-mode))

(setq-default tab-width 2)

(custom-set-variables
;; custom-set-variables was added by Custom.
;; If you edit it by hand, you could mess it up, so be careful.
;; Your init file should contain only one such instance.
;; If there is more than one, they won't work right.
'(ansi-color-faces-vector
[default default default italic underline success warning error])
'(ansi-color-names-vector
["#242424" "#e5786d" "#95e454" "#cae682" "#8ac6f2" "#333366" "#ccaa8f" "#f6f3e8"])
'(custom-enabled-themes (quote (tsdh-dark))))
(custom-set-faces
;; custom-set-faces was added by Custom.
;; If you edit it by hand, you could mess it up, so be careful.
;; Your init file should contain only one such instance.
;; If there is more than one, they won't work right.
)

(setq scheme-root "Applications/MIT-Scheme.app/Contents/Resources")

;; (setq scheme-program-name
;;       (concat
;;        scheme-root "/mit-scheme "
;;        "--library " scheme-root " "
;;        "--band " scheme-root "/all.com "
;;        "-heap 10000"))

(setq scheme-program-name "/Applications/MIT-Scheme.app/Contents/Resources/mit-scheme")

(add-to-list 'load-path "~/git/rainbow-delimiters")
(require 'rainbow-delimiters)
(add-hook 'prog-mode-hook 'rainbow-delimiters-mode)

;;; I prefer cmd key for meta
(setq mac-option-key-is-meta nil
      mac-command-key-is-meta t
      mac-command-modifier 'meta
      mac-option-modifier 'none)
(set-keyboard-coding-system nil)

;;; Enable Ido mode by default
(setq ido-enable-flex-matching t)
(setq ido-everywhere t)
(setq ido-create-new-buffer 'always)
(ido-mode 1)

;;; Setup mit-scheme config
(setq scheme-program-name "/usr/local/bin/mit-scheme")
